pkg/addon/agent/controller/token.go

```diff
@@ -33,37 +33,57 @@ type TokenReconciler struct {
 	SpokeNativeClient kubernetes.Interface
 	SpokeClientConfig *rest.Config
 	SpokeNamespace    string
+	ClusterName       string
+	SpokeCache        cache.Cache
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *TokenReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&authv1alpha1.ManagedServiceAccount{}).
-		Watches(&source.Kind{Type: &corev1.Secret{}}, event.NewSecretEventHandler()).
+		Watches(
+			&source.Kind{
+				Type: &corev1.Secret{},
+			},
+			event.NewSecretEventHandler(),
+		).
+		Watches(
+			source.NewKindWithCache(
+				&corev1.ServiceAccount{},
+				r.SpokeCache,
+			),
+			event.NewServiceAccountEventHandler(r.ClusterName),
+		).
 		Complete(r)
 }
 
 func (r *TokenReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
 	logger := log.FromContext(ctx)
 	logger.Info("Start reconciling")
 	managed := &authv1alpha1.ManagedServiceAccount{}
+
 	if err := r.Cache.Get(ctx, request.NamespacedName, managed); err != nil {
 		if !apierrors.IsNotFound(err) {
-			return reconcile.Result{}, errors.Wrapf(err, "no such managed service account")
+			//fail to get managed-serviceaccount, requeue
+			return reconcile.Result{}, errors.Wrapf(err, "fail to get managed serviceaccount")
+		}
+
+		sai := r.SpokeNativeClient.CoreV1().ServiceAccounts(r.SpokeNamespace)
+		if err := sai.Delete(ctx, request.Name, metav1.DeleteOptions{}); err != nil {
```

Comment:
the ideal way of synchronizing the deletion of hub-object and spoke-object is using the finalizer. e.g. we attach a `authentication.open-cluster-management.io/service-account-cleanup`?

Reply:
> any specific reason for needing a finalizer to facilitate this behavior?

the deletion call can fail unexpectedly sometimes, i think if the deletion of the managed SA succeeded but the deletion of local SA failed. the leftover local SA will not be cleaned up in the future? but i see we added a event listener watching the local SA, so at lease they can be periodically cleaned up.

> potentially introduce edge case that will cause serviceaccount deletion to "wedge"

yea i did run into the wedging deletion problem before. i think ideally the graceful deletion flow should be managed by the addon-framework in the future.